compiler/rustc_mir_transform/src/inline.rs

```diff
@@ -964,3 +983,41 @@ impl<'a, 'tcx> MutVisitor<'tcx> for Integrator<'a, 'tcx> {
         }
     }
 }
+
+crate fn mir_inliner_callers_count_map<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    _: (),
+) -> Lrc<FxHashMap<DefId, usize>> {
+    let mut callers_count_map: FxHashMap<DefId, usize> = Default::default();
+
+    // Iterate over all body owners of current crate
+    for did in tcx
+        .hir()
+        .body_owners()
+        // We are interested only in functions or closures
+        .filter(|did| {
+            tcx.hir().body_owner_kind(tcx.hir().local_def_id_to_hir_id(*did)).is_fn_or_closure()
+        })
+    {
+        // Get unomptimized body to avoid query cycles.
+        let body = tcx
+            .mir_drops_elaborated_and_const_checked(ty::WithOptConstParam::unknown(did))
+            .borrow();
+
+        let param_env = tcx.param_env_reveal_all_normalized(did);
+
+        body.basic_blocks()
+            .iter_enumerated()
+            .filter_map(|(bb, bb_data)| {
+                // Try to find call site in this basic block.
+                CallSite::resolve_callsite(tcx, param_env, &body, bb, bb_data)
+            })
+            // Get function def_id which are we calling.
+            .map(|callsite| callsite.callee.def_id())
+            .for_each(|did| {
+                *callers_count_map.entry(did).or_insert(0) += 1;
+            });
```

Comment:
Could you rewrite this combinator chain as a simple `for` loop? I find it a little hard to follow.